### PR TITLE
smlnj: fix x86_64-darwin build

### PR DESCRIPTION
**Pull request by simonchatts**
###### Motivation for this change

Fix the `x86_64-darwin` build for Standard ML of New Jersey. I've confirmed that `x86_64-linux` isn't affected.

This entails two small changes: correcting the 32/64-bit input to the project's build scripts, and locating standard C headers in the nix store.

I thought the 32/64-bit fix might also enable `aarch64-linux`, but the project is sufficiently old that it needs additional updates to support ARM at all. Since the [latest README](http://www.smlnj.org/dist/working/110.99.1/110.99.1-README.html) says they are actively working on native Apple M1 support, I haven't gone any further here.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by xworld21** [comment]
Three packages (`jonprl`, `celf`, `twelf`) use `smlnjBootstrap` on Darwin, should those be updated to use `smlnj`?

File: pkgs/development/compilers/smlnj/default.nix [hunk 1]
```diff
@@ -1,11 +1,9 @@
-{ lib, stdenv, fetchurl }:
+{ lib, stdenv, fetchurl, pkgs }:
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/compilers/smlnj/default.nix:1]: Neither pkgs or darwin is allowed as an input.

File: pkgs/development/compilers/smlnj/default.nix [hunk 2]
```diff
@@ -53,6 +51,10 @@ in stdenv.mkDerivation {
     sed -i '/PATH=/d' config/_arch-n-opsys base/runtime/config/gen-posix-names.sh
     echo SRCARCHIVEURL="file:/$TMP" > config/srcarchiveurl
     patch --verbose config/_heap2exec ${./heap2exec.diff}
+  '' + lib.optionalString stdenv.isDarwin ''
+    # Locate standard headers like <unistd.h>
+    substituteInPlace base/runtime/config/gen-posix-names.sh \
+      --replace "\$SDK_PATH/usr" "${pkgs.darwin.Libsystem}"
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/smlnj/default.nix:57]:
```suggestion
      --replace "\$SDK_PATH/usr" "${Libsystem}"
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by simonchatts** [comment]
Thanks for the quick comments everyone! I've fixed the inputs as per @SuperSandro2000's suggestion.

I've also checked out @xworld21's question about downstream dependents, and think they are right. On both `x86_64-darwin` and `x86_64-linux`, unconditionally moving `twelf` and `celf` to the new `smlnj` seems to work fine. `jonprl` is [already broken and linux-only](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/science/logic/jonprl/default.nix#L32-L33), but I see no obvious reason why, if these were fixed, why it should work with `smlnjBootstrap` but not the new `smlnj`. So I've done this too.

File: pkgs/development/compilers/smlnj/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonchatts) [reply pkgs/development/compilers/smlnj/default.nix]: Thanks - resolved as below.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -12055,8 +12055,9 @@ in
 
   shmig = callPackage ../development/tools/database/shmig { };
 
+  # smlnjBootstrap should be redundant, now that smlnj works on Darwin natively
   smlnjBootstrap = callPackage ../development/compilers/smlnj/bootstrap.nix { };
-  smlnj = callPackage ../development/compilers/smlnj { };
+  smlnj = callPackage ../development/compilers/smlnj { inherit (darwin) Libsystem; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:12060]:
```suggestion
  smlnj = callPackage ../development/compilers/smlnj {
    inherit (darwin) Libsystem;
  };
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (simonchatts) [reply pkgs/top-level/all-packages.nix:12060]: Thanks - fixed!

**Comment by xworld21** [comment]
It looks good to me. For good measure, let me cc @jwiegley (twelf and smlnjBootstrap maintainer) and @bcdarwin (celf maintainer) since it affects their packages.

**Pull request merged**